### ISSUE TITLE
Suggested cleanups

### ISSUE THREAD

**Issue by davidweichiang** [diprism/perpl#79]
None of these suggestions changes the behavior of the compiler, but might simplify the code:

- [x] #84

- [x] Currently a datatype's type parameters are stored in the `Env` or `Ctxt`, while a define's type parameters are stored in `Scheme`/`Forall`. Can we treat them both the same way? (Not sure which is better.)

- [x] #90

- [ ] There's a lot of duplicate code between `UsProgs`, `SProgs`, and `Progs`. Would it be possible to merge them?

**Comment by davidweichiang** [comment]
@ccshan Are you still interested in pursuing https://github.com/diprism/perpl/tree/type-params? Do you want to make it into a PR so it can be discussed?

**Comment by ccshan** [comment]
> @ccshan Are you still interested in pursuing https://github.com/diprism/perpl/tree/type-params? Do you want to make it into a PR so it can be discussed?

Yes, and I can probably do it faster and easier the second time, but it seems like we should avoid having multiple PRs that each make many changes textually speaking?

**Comment by davidweichiang** [comment]
Here's what Prog and Ctxt currently look like:

x = identifier, tp = type, tm = term, p = parameter name, ptp = parameter type, cs = constructors, rtp = return type, ps = type parameters

|   | User-level | Scheme-ified | Monomorphized | Argified
|---|---|---|---|---
| define | UsProgFun x tp tm | SProgFun x tgs ps tp tm | ProgFun x [] tm rtp | ProgFun x [(p, ptp)] tm rtp
| extern | UsProgExtern x tp | SProgExtern x [] rtp | ProgExtern x [] rtp | ProgExtern x [ptp] rtp
| data | UsProgData x ps cs | SProgData x tgs ps cs | ProgData x cs | ProgData x cs

| | User-level | Scheme-ified | Elaborated
-- | -- | -- | --
| local var | DefTerm (DefLocal tp) | DefTerm (DefLocal tp) | DefTerm (DefLocal tp)
| define | DefTerm (DefGlobal [] [] tp) | DefTerm (DefGlobal tgs ps tp) | DefTerm (DefGlobal [] [] tp)
| extern | DefTerm (DefGlobal [] [] tp) | DefTerm (DefGlobal tgs* [] tp) | DefTerm (DefGlobal [] [] tp)
| constructor | DefTerm (DefCtor ps [] tp) | DefSTerm (DefCtor tgs ps tp) | DefTerm (DefCtor [] [] tp)
| datatype | DefType (DefData ps [] cs) | DefType (DefData tgs ps cs) | DefType (DefData [] cs)
| type var | ** | ** | **

*An extern can only have tags if it has a type parameter that is ignored. In practice I think we just store [] for the tags.
**We currently just don't store type variables in Ctxt.

**Comment by davidweichiang** [comment]
I'm editing the above charts to keep them up to date, instead of making new ones.

**Comment by davidweichiang** [comment]
Would it be possible/advisable to merge ProgFun/ProgExtern with DefTerm, ProgData with DefType? That is, a Progs would basically be a Ctxt together with an end Term?

**Comment by ccshan** [comment]
> Would it be possible/advisable to merge ProgFun/ProgExtern with DefTerm, ProgData with DefType? That is, a Progs would basically be a Ctxt together with an end Term?

That would be nice but I think the merger should happen in the opposite direction: Instead of worrying about the consistency among the `DefCtor`s and `DefData`s of a `Ctxt` when transformed, we would have `type Ctxt = [Prog]`, and `ctxtLookupTerm` would look through all 3 kinds of `Prog`s because the given `Var` might be a constructor name. But I understand that `ctxtDefLocal` is not representable as a `Prog`, and looking up a `Var` in the `Map` that is `Ctxt` is a little faster than in the list that is `[Prog]`, so maybe a merger is not so simple.

**Comment by davidweichiang** [comment]
- [ ] It would be nice to eliminate `NoTp`.

**Comment by davidweichiang** [comment]
> * [ ]  It would be nice to eliminate `NoTp`.

I'm not entirely sure what I had in mind when I wrote this, but perhaps it was that in the parser, terms that currently receive `NoTp` should receive a fresh `TpVar` instead?
